File: pkg/components/interactive/tabs/tabs.go
// Package tabs provides an accessible tabs component using ARIA roles and data attributes.
// The initial active state is rendered server-side (SSR-ready, works without JS).
// app.js initTabs() attaches click handlers for switching panels at runtime.
package tabs

import (
	g "maragu.dev/gomponents"
	h "maragu.dev/gomponents/html"
)

// Root renders the root tabs container. defaultTab sets the initially active tab value.
// The value is stored in data-tabs so initTabs() in app.js can read it.
func Root(defaultTab string, children ...g.Node) g.Node {
	return h.Div(
		g.Attr("data-tabs", defaultTab),
		h.Class("w-full"),
		g.Group(children),
	)
}

// List renders the tab button bar.
func List(children ...g.Node) g.Node {
	return h.Div(
		h.Role("tablist"),
		h.Class("inline-flex h-9 items-center justify-center rounded-lg bg-muted p-1 text-muted-foreground"),
		g.Group(children),
	)
}

// Trigger renders a single tab button. isDefault marks the initially active tab,
// which is reflected via aria-selected and active styling without requiring JS.
func Trigger(value string, isDefault bool, children ...g.Node) g.Node {
	cls := "inline-flex items-center justify-center whitespace-nowrap rounded-md px-3 py-1 text-sm font-medium transition-all focus-visible:outline-none focus-visible:ring-2 disabled:pointer-events-none disabled:opacity-50"
	ariaSelected := "false"
	if isDefault {
		ariaSelected = "true"
	}
	nodes := []g.Node{
		h.Type("button"),
		h.Role("tab"),
		g.Attr("data-tab", value),
		g.Attr("aria-selected", ariaSelected),
		h.Class(cls),
	}
	if isDefault {
		nodes = append(nodes, h.Class("bg-background text-foreground shadow"))
	}
	nodes = append(nodes, g.Group(children))
	return h.Button(nodes...)
}

// Content renders the panel for a specific tab value. isDefault controls
// whether the panel is visible on initial render (hidden attr set when false).
func Content(value string, isDefault bool, children ...g.Node) g.Node {
	nodes := []g.Node{
		h.Role("tabpanel"),
		g.Attr("data-tab", value),
		h.Class("mt-2"),
	}
	if !isDefault {
		nodes = append(nodes, g.Attr("hidden", ""))
	}
	nodes = append(nodes, g.Group(children))
	return h.Div(nodes...)
}

File: static/js/components/tabs.js
import { data, findAll, registerUpgrade, toggleClass } from '../lib/dom.js';

var activeClasses = ['bg-background', 'text-foreground', 'shadow'];

function activateTab(container, trigger, focus) {
  var triggers = findAll(container, '[role="tab"]');
  var panels = findAll(container, '[role="tabpanel"]');
  var activeTab = data(trigger, 'tab');

  triggers.forEach(function (candidate) {
    var active = candidate === trigger;
    candidate.setAttribute('aria-selected', active ? 'true' : 'false');
    candidate.setAttribute('tabindex', active ? '0' : '-1');
    activeClasses.forEach(function (className) {
      toggleClass(candidate, className, active);
    });
  });

  panels.forEach(function (panel) {
    panel.hidden = data(panel, 'tab') !== activeTab;
  });

  if (focus) {
    trigger.focus();
  }
}

function mount(container) {
  if (container.dataset.tabsInitialized === 'true') {
    return;
  }

  var triggers = findAll(container, '[role="tab"]');
  if (!triggers.length) {
    return;
  }

  container.dataset.tabsInitialized = 'true';
  var selected = triggers.find(function (trigger) {
    return trigger.getAttribute('aria-selected') === 'true';
  }) || triggers[0];

  activateTab(container, selected, false);

  triggers.forEach(function (trigger, index) {
    trigger.addEventListener('click', function () {
      activateTab(container, trigger, false);
    });

    trigger.addEventListener('keydown', function (event) {
      var nextIndex = index;
      if (event.key === 'ArrowRight') {
        nextIndex = (index + 1) % triggers.length;
      } else if (event.key === 'ArrowLeft') {
        nextIndex = (index - 1 + triggers.length) % triggers.length;
      } else if (event.key === 'Home') {
        nextIndex = 0;
      } else if (event.key === 'End') {
        nextIndex = triggers.length - 1;
      } else {
        return;
      }

      event.preventDefault();
      activateTab(container, triggers[nextIndex], true);
    });
  });
}

registerUpgrade(function (root) {
  findAll(root, '[data-tabs]').forEach(mount);
});
